colossalai/pipeline/p2p.py

```diff
@@ -131,6 +146,224 @@ def _broadcast_object_list(
             object_list[i] = unpickle_object
 
 
+def check_device(group):
+    is_nccl_backend = check_for_nccl_backend(group)
+    current_device = None
+
+    current_device = torch.device("cpu")
+    if is_nccl_backend:
+        current_device = torch.device("cuda", torch.cuda.current_device())
+    return current_device, is_nccl_backend
+
+
+def filling_ops_queue(obj, comm_op, comm_rank, ops_queue, group):
+    if isinstance(obj, torch.Tensor):
+        op_to_add = dist.P2POp(comm_op, obj, comm_rank, group)
+        ops_queue.append(op_to_add)
+    else:
+        for tensor_to_comm in obj:
+            op_to_add = dist.P2POp(comm_op, tensor_to_comm, comm_rank, group)
+            ops_queue.append(op_to_add)
+
+
+def create_recv_buffer(tensor_metadata, current_device):
+    if tensor_metadata[0] == 0:
+        tensor_shape, tensor_dtype, tensor_requires_grad = tensor_metadata[1]
+        tensor_recv = torch.empty(
+            tensor_shape, requires_grad=tensor_requires_grad, device=current_device, dtype=tensor_dtype)
+        return tensor_recv
+    elif tensor_metadata[0] == 1 or tensor_metadata[0] == 2:
+        buffer_recv = []
+        for tensor_data in tensor_metadata[1:]:
+            tensor_shape, tensor_dtype, tensor_requires_grad = tensor_data[-3:]
+            tensor_recv = torch.empty(
+                tensor_shape, requires_grad=tensor_requires_grad, device=current_device, dtype=tensor_dtype)
+            buffer_recv.append(tensor_recv)
+        return buffer_recv
+
+
+def _batch_send_recv_tensor(send_tensor_list, recv_tensor_metadata, send_dst, recv_src, send_group, recv_group, current_device):
+    buffer_recv = None
+    if recv_tensor_metadata is not None:
+        buffer_recv = create_recv_buffer(recv_tensor_metadata, current_device)
+
+    ops = []
+
+    if send_dst is not None:
+        filling_ops_queue(send_tensor_list, dist.isend, send_dst, ops, send_group)
+
+    if recv_src is not None:
+        assert buffer_recv is not None
+        filling_ops_queue(buffer_recv, dist.irecv, recv_src, ops, recv_group)
+
+    if len(ops) > 0:
+        reqs = dist.batch_isend_irecv(ops)
+        for req in reqs:
+            req.wait()
+    torch.cuda.synchronize()
```

Comment:
I didn't find another stream. I think the synchronization is not necessary, right?

Reply:
megatron did it here
https://github.com/microsoft/Megatron-DeepSpeed/blob/ef13d099c2a1609225a4ce4c1a1753cc76dd90a1/megatron/p2p_communication.py#L111-L112

I can remove it if it is handled properly

Reply:
Removed synchronization

Reply:
Pytorch's documentation does not sychronize. In case there is potential error, could you leave a comment here? If we get error in further training, we can modify it.

Reply:
Added